File: frontend/js/views/disability_insurance_needs.js
// Disability-insurance needs — monthly benefit gap net of group LTD, vs
// expenses, via /calc/disability-insurance-needs. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const x = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '×');
const VIEW = 'disability-insurance-needs';
let lastReport = null;
let lastBody = null;

export async function renderDisabilityInsuranceNeeds(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.disab.h1.title">// DISABILITY INSURANCE NEEDS</span></h1>
        <p class="muted small" data-i18n="view.disab.hint.intro">
            The monthly benefit that replaces lost income if you can't work, net of any employer
            long-term-disability coverage. Policies typically replace ~60% of income (benefits are
            tax-free on premiums you pay). The gap is the extra benefit to buy, checked against your
            essential expenses. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.disab.h2.inputs">Your situation</h2>
            <form id="disab-form" class="inline-form">
                <label><span data-i18n="view.disab.label.income">Annual income ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_income_usd" value="90000" required></label>
                <label><span data-i18n="view.disab.label.replacement">Replacement (% of income)</span>
                    <input type="number" step="1" min="0" max="100" name="replacement_pct" value="60" required></label>
                <label><span data-i18n="view.disab.label.existing">Existing LTD / mo ($)</span>
                    <input type="number" step="0.01" min="0" name="existing_coverage_monthly_usd" value="2000"></label>
                <label><span data-i18n="view.disab.label.expenses">Monthly expenses ($)</span>
                    <input type="number" step="0.01" min="0" name="monthly_expenses_usd" value="4000"></label>
            </form>
            <div id="disab-tools" class="ce-toolbar"></div>
            <button type="button" id="disab-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="disab-sens" class="ce-sens"></div>
        </div>
        <div id="disab-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#disab-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            annual_income_usd: Number(fd.get('annual_income_usd')) || 0,
            replacement_pct: Number(fd.get('replacement_pct')) || 0,
            existing_coverage_monthly_usd: Number(fd.get('existing_coverage_monthly_usd')) || 0,
            monthly_expenses_usd: Number(fd.get('monthly_expenses_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcDisabilityInsuranceNeeds(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.disab.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#disab-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'disability-insurance-needs.csv' });
    mount.querySelector('#disab-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['monthly_gap_usd', r.monthly_gap_usd],
        ['target_monthly_benefit_usd', r.target_monthly_benefit_usd],
        ['expense_coverage_ratio', r.expense_coverage_ratio],
        ['monthly_income_usd', r.monthly_income_usd],
        ['annual_gap_usd', r.annual_gap_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#disab-result');
    const gapCls = r.monthly_gap_usd > 0 ? 'neg' : 'pos';
    const coverCls = r.covers_expenses ? 'pos' : 'neg';
    // Line chart: monthly benefit gap as replacement % sweeps 40% -> 80%.
    const xs = enh.linspace(40, 80, 13);
    const pts = await Promise.all(xs.map(async (p) => {
        const rr = await api.calcDisabilityInsuranceNeeds({ ...body, replacement_pct: p });
        return { x: p, y: rr ? rr.monthly_gap_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'replace %', ylabel: 'gap $/mo' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.disab.h2.result">The coverage gap</h2>
            <div class="cards">
                <div class="card ${gapCls}"><div class="label" data-i18n="view.disab.card.gap">Monthly gap</div>
                    <div class="value ${gapCls}">${money(r.monthly_gap_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.disab.card.target">Target benefit / mo</div>
                    <div class="value">${money(r.target_monthly_benefit_usd)}</div></div>
                <div class="card ${coverCls}"><div class="label" data-i18n="view.disab.card.cover">Covers expenses</div>
                    <div class="value ${coverCls}">${x(r.expense_coverage_ratio)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.disab.row.income">Monthly income</td><td>${money(r.monthly_income_usd)}</td></tr>
                    <tr><td data-i18n="view.disab.row.target">Target monthly benefit</td><td>${money(r.target_monthly_benefit_usd)}</td></tr>
                    <tr><td data-i18n="view.disab.row.annual">Annual benefit gap</td><td>${money(r.annual_gap_usd)}</td></tr>
                    <tr class="emph ${gapCls}"><td data-i18n="view.disab.row.gap">Monthly benefit gap</td><td>${money(r.monthly_gap_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#disab-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: replacement % 40 -> 80; y: existing LTD 0 -> 4000/mo. Output: monthly gap (lower better -> negate).
    const xVals = enh.linspace(40, 80, 5);
    const yVals = enh.linspace(0, 4000, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'replacement_pct', yKey: 'existing_coverage_monthly_usd', xVals, yVals, compute: (b) => api.calcDisabilityInsuranceNeeds(b), pick: (r) => (r ? r.monthly_gap_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : '$' + Math.round(-v)), xfmt: (v) => v.toFixed(0) + '%', yfmt: (v) => '$' + Math.round(v), xName: t('view.disab.label.replacement') || 'Replace', yName: t('view.disab.label.existing') || 'LTD' });
}

File: crates/traderview-core/src/disability_insurance_needs.rs
//! Disability-insurance needs — the monthly benefit that replaces lost income
//! if you can't work, net of any employer long-term-disability (LTD) coverage.
//!
//! Disability policies replace a fraction of income (commonly ~60%, since
//! benefits on premiums you pay are tax-free). The gap is the extra monthly
//! benefit to buy beyond group coverage, checked against actual expenses.
//!
//! ```text
//! target benefit = replacement % × gross monthly income
//! monthly gap    = target − existing LTD coverage   (floored at 0)
//! ```

use serde::{Deserialize, Serialize};

fn d_replacement() -> f64 {
    60.0
}

#[derive(Debug, Clone, Deserialize)]
pub struct DisabilityInput {
    pub annual_income_usd: f64,
    /// Share of income to replace, percent (≈60% is typical).
    #[serde(default = "d_replacement")]
    pub replacement_pct: f64,
    /// Existing employer/group LTD benefit, monthly.
    #[serde(default)]
    pub existing_coverage_monthly_usd: f64,
    /// Essential monthly expenses, to check coverage adequacy.
    #[serde(default)]
    pub monthly_expenses_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DisabilityResult {
    pub monthly_income_usd: f64,
    /// Target monthly benefit (replacement % of income).
    pub target_monthly_benefit_usd: f64,
    /// Additional monthly benefit to buy beyond existing coverage.
    pub monthly_gap_usd: f64,
    /// Annualized gap.
    pub annual_gap_usd: f64,
    /// Target benefit ÷ monthly expenses; `None` if no expenses given.
    pub expense_coverage_ratio: Option<f64>,
    /// Whether the target benefit covers essential expenses.
    pub covers_expenses: bool,
}

pub fn analyze(input: &DisabilityInput) -> DisabilityResult {
    let monthly_income = input.annual_income_usd / 12.0;
    let target = input.replacement_pct / 100.0 * monthly_income;
    let gap = (target - input.existing_coverage_monthly_usd).max(0.0);

    let (ratio, covers) = if input.monthly_expenses_usd > 0.0 {
        let r = target / input.monthly_expenses_usd;
        (Some(r), target >= input.monthly_expenses_usd)
    } else {
        (None, true)
    };

    DisabilityResult {
        monthly_income_usd: monthly_income,
        target_monthly_benefit_usd: target,
        monthly_gap_usd: gap,
        annual_gap_usd: gap * 12.0,
        expense_coverage_ratio: ratio,
        covers_expenses: covers,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn base() -> DisabilityInput {
        DisabilityInput {
            annual_income_usd: 90_000.0,
            replacement_pct: 60.0,
            existing_coverage_monthly_usd: 2_000.0,
            monthly_expenses_usd: 4_000.0,
        }
    }

    #[test]
    fn monthly_income_and_target() {
        let r = analyze(&base());
        assert!(close(r.monthly_income_usd, 7_500.0));
        // 60% × 7,500 = 4,500.
        assert!(close(r.target_monthly_benefit_usd, 4_500.0));
    }

    #[test]
    fn monthly_gap() {
        // 4,500 target − 2,000 existing = 2,500.
        let r = analyze(&base());
        assert!(close(r.monthly_gap_usd, 2_500.0));
        assert!(close(r.annual_gap_usd, 30_000.0));
    }

    #[test]
    fn expense_coverage() {
        let r = analyze(&base());
        // 4,500 / 4,000 = 1.125 — covers expenses.
        assert!(close(r.expense_coverage_ratio.unwrap(), 1.125));
        assert!(r.covers_expenses);
    }

    #[test]
    fn does_not_cover_high_expenses() {
        let r = analyze(&DisabilityInput {
            monthly_expenses_usd: 6_000.0,
            ..base()
        });
        assert!(!r.covers_expenses);
        assert!(r.expense_coverage_ratio.unwrap() < 1.0);
    }

    #[test]
    fn ample_existing_coverage_zero_gap() {
        let r = analyze(&DisabilityInput {
            existing_coverage_monthly_usd: 5_000.0,
            ..base()
        });
        assert!(close(r.monthly_gap_usd, 0.0));
    }

    #[test]
    fn higher_replacement_raises_target() {
        let low = analyze(&base());
        let high = analyze(&DisabilityInput {
            replacement_pct: 80.0,
            ..base()
        });
        assert!(high.target_monthly_benefit_usd > low.target_monthly_benefit_usd);
    }

    #[test]
    fn no_expenses_assumes_covered() {
        let r = analyze(&DisabilityInput {
            monthly_expenses_usd: 0.0,
            ..base()
        });
        assert!(r.expense_coverage_ratio.is_none());
        assert!(r.covers_expenses);
    }

    #[test]
    fn existing_coverage_reduces_gap() {
        let less = analyze(&DisabilityInput {
            existing_coverage_monthly_usd: 1_000.0,
            ..base()
        });
        let more = analyze(&DisabilityInput {
            existing_coverage_monthly_usd: 3_000.0,
            ..base()
        });
        assert!(more.monthly_gap_usd < less.monthly_gap_usd);
    }
}
